### PR TITLE
Bump hcsshim to b3f49c06ffaeef24d09c6c08ec8ec8425a0303e2

### LINKED ISSUE #5: Docker won't build container.
Hello,
i am new to docker and i have an issue. The script doesn't do anything and i get this error:

Starting Docker container to build Audacity with ASIO support
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: hcsshim::PrepareLayer - failed failed in Win32: Unzulässige Funktion. (0x1).

Any idea what i can do?

WinVer: 1903 - 18362.295
Docker: 2.1.0.3

### LINKED ISSUE #3523: Error when running build with storage-opts set
Hi,

when I change the storage-opts of the docker daemon like specified in the instructions, I get the following error during the build:
hcsshim::PrepareLayer failed in Win32: The parameter is incorrect. (0x57) path=C:\ProgramData\Docker\windowsfilter\57da5c77f4290bce34c408cf013675985afdfd3ad2b9b5c578b2b93521c96559

This bug seems to have been introduced in Windows 1903 (see https://github.com/moby/moby/issues/36831), is there a workaround available for this?

Thanks,

Joerg

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b91a4a43-5353-97cf-7532-7644aa3b497f
* Version Independent ID: 858d9a5c-72f3-6a85-9270-0a0dd9e0438b
* Content: [Install Visual Studio Build Tools into a container](https://docs.microsoft.com/en-us/visualstudio/install/build-tools-container?view=vs-2019#feedback)
* Content Source: [docs/install/build-tools-container.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/install/build-tools-container.md)
* Product: **visual-studio-windows**
* Technology: **vs-installation**
* GitHub Login: @heaths
* Microsoft Alias: **tglee**

### LINKED ISSUE #36831: hcsshim::PrepareLayer failed in Win32: The parameter is incorrect
Is this a bug or am I missing something?
When I run `docker-compose build` I get this error:

```
Building webapplication1
Step 1/17 : FROM microsoft/aspnetcore:2.0-nanoserver-sac2016 AS base
 ---> 7f9d7dc1860e
Step 2/17 : WORKDIR /app
 ---> Using cache
 ---> dd80e4a90ee4
Step 3/17 : EXPOSE 80
 ---> Using cache
 ---> 61d4c528d49e
Step 4/17 : FROM microsoft/aspnetcore-build:2.0-nanoserver-sac2016 AS build
 ---> 1f96313b74e3
Step 5/17 : WORKDIR /src
ERROR: Service 'webapplication1' failed to build: hcsshim::PrepareLayer failed in Win32: The parameter is incorrect. (0x57) layerId=a96bac7ce3ed61f622fc7ec84113221af4ee9a24d792b8a0c7785c6b316c3512 flavour=1
```

docker version:
```
Client:
 Version:       18.03.0-ce
 API version:   1.37
 Go version:    go1.9.4
 Git commit:    0520e24
 Built: Wed Mar 21 23:06:28 2018
 OS/Arch:       windows/amd64
 Experimental:  false
 Orchestrator:  swarm

Server:
 Engine:
  Version:      18.03.0-ce
  API version:  1.37 (minimum version 1.24)
  Go version:   go1.9.4
  Git commit:   0520e24
  Built:        Wed Mar 21 23:21:06 2018
  OS/Arch:      windows/amd64
  Experimental: false
```
docker info:
```
Containers: 4
 Running: 1
 Paused: 0
 Stopped: 3
Images: 4
Server Version: 18.03.0-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 14393 (14393.2125.amd64fre.rs1_release.180301-2139)
Operating System: Windows 10 Enterprise
OSType: windows
Architecture: x86_64
CPUs: 8
Total Memory: 31.85GiB
Name: myComputerName
ID: JBIL:ZT3U:37LT:YMJA:5LDM:PGM7:L7PN:DDVF:HSDN:VI6S:TS72:A67K
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 25
 System Time: 2018-04-11T10:56:44.0030666-04:00
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

### LINKED ISSUE #39524: Docker build windows 19.03 --storage-opt size>20G
**Description**
Unable to build windows images larger then 20 GB.
Am able to run images with --storage-opt size>20G, either with docker run --storage-opt or dockerd --register-service --storage-opt

**Steps to reproduce the issue:**
1. Windows 10 19.03 
Version	10.0.18362 build 18362
2. 
Dockerfile:
FROM mcr.microsoft.com/windows/servercore:ltsc2019
WORKDIR /image/
3. net stop docker;
dockerd --unregister-service;
dockerd --register-service --storage-opt "size=22G";
net start docker;
docker build -t docker-test:latest --no-cache . ; 
docker run --rm docker-test:latest powershell -c get-psdrive;

**Describe the results you received:**
With storage-opt <20G docker will build and run. Both will not listen to storage-opt size<=20G and run disk at 20G anyways.
With storage-opt >20G docker won't build with error: "hcsshim::PrepareLayer failed in Win32: The parameter is incorrect. (0x57)"
But will run an image with disk >20G as set by dockerd

**Describe the results you expected:**
Docker build to listen to storage-opt as docker run does.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:                                                   
 Version:           18.09.6                               
 API version:       1.39                                  
 Go version:        go1.10.8                              
 Git commit:        c89750f8                              
 Built:             06/09/2019 17:10:42                   
 OS/Arch:           windows/amd64                         
 Experimental:      false                                 
                                                          
Server: Docker Engine - Community                         
 Engine:                                                  
  Version:          18.09.2                               
  API version:      1.39 (minimum version 1.24)           
  Go version:       go1.10.6                              
  Git commit:       6247962                               
  Built:            Sun Feb 10 04:28:48 2019              
  OS/Arch:          windows/amd64                         
  Experimental:     false                                 
```

**Output of `docker info`:**

```
Containers: 1                                                                                     
 Running: 0                                                                                       
 Paused: 0                                                                                        
 Stopped: 1                                                                                       
Images: 27                                                                                        
Server Version: 18.09.2                                                                           
Storage Driver: windowsfilter                                                                     
 Windows:                                                                                         
Logging Driver: json-file                                                                         
Plugins:                                                                                          
 Volume: local                                                                                    
 Network: ics l2bridge l2tunnel nat null overlay transparent                                      
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog                       
Swarm: inactive                                                                                   
Default Isolation: hyperv                                                                         
Kernel Version: 10.0 18362 (18362.1.amd64fre.19h1_release.190318-1202)                            
Operating System: Windows 10 Enterprise Version 1903 (OS Build 18362.207)                         
OSType: windows                                                                                   
Architecture: x86_64                                                                              
CPUs: 4                                                                                           
Total Memory: 31.87GiB                                                                            
Name: COLBTBTNH2                                                                                  
ID: 6BXE:H65O:6ZHQ:4SCZ:T662:RSVA:LD6X:5SKV:6T43:6ZP6:WAL5:VYOK                                   
Docker Root Dir: C:\ProgramData\Docker                                                            
Debug Mode (client): false                                                                        
Debug Mode (server): true                                                                         
 File Descriptors: -1                                                                             
 Goroutines: 23                                                                                   
 System Time: 2019-07-15T15:43:39.7056454+02:00                                                   
 EventsListeners: 0                                                                               
Registry: https://index.docker.io/v1/                                                             
Labels:                                                                                           
Experimental: false                                                                               
Insecure Registries:                                                                              
 127.0.0.0/8                                                                                      
Live Restore Enabled: false                                                                       
Product License: Community Engine                                                                 
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Localmachine. Tested briefly on windows servercore with containers on Azure but was not able to get net start docker to work.

### PR DESCRIPTION
**Pull request by vikramhh**
Opening this PR to carry https://github.com/moby/moby/pull/40128 in two different commits in the correct order.

~rebased on top of https://github.com/moby/moby/pull/40154~

- [x] depends on https://github.com/moby/moby/pull/40154 bump containerd/cgroups 5fbad35c2a7e855762d3c60f2e474ffcad0d470a

Among other things, this is required to pull in
https://github.com/microsoft/hcsshim/pull/718 which should
take care of multiple issues reported to us.

full diff: https://github.com/microsoft/hcsshim/compare/2226e083fc390003ae5aa8325c3c92789afa0e7a...6c7177eae8be632af2e15e44b62d69ab18389ddb

includes:

- https://github.com/microsoft/hcsshim/pull/718 wclayer: Work around Windows bug when expanding sandbox size
    - fixes https://github.com/microsoft/hcsshim/issues/708 Windows Host Compute Service bug breaks docker (and other) sandboxes bigger than 20G on Windows 1903
    - fixes https://github.com/microsoft/hcsshim/issues/624The hcsshim on Windows 10 1903 always fails to build Docker image
    - fixes/addresses https://github.com/docker/for-win/issues/3884 An error occurred while attempting to build Docker image (especially this comment and the next comments after: https://github.com/docker/for-win/issues/3884#issuecomment-498939672)
    - fixes/addresses https://github.com/docker/for-win/issues/4100 Windows 1903 fails when storage-opt used 
    - fixes https://github.com/moby/moby/issues/36831 hcsshim::PrepareLayer failed in Win32: The parameter is incorrect (https://github.com/moby/moby/issues/36831#issuecomment-498612392)
    - fixes https://github.com/Stannieman/audacity-with-asio-builder/issues/5 Docker won't build container
    - fixes https://github.com/MicrosoftDocs/visualstudio-docs/issues/3523 Error when running build with storage-opts set
    - fixes https://github.com/moby/moby/issues/39524 Docker build windows 19.03 --storage-opt size>20G

Note that this is a temporary workaround for a bug in the platform, and will be reverted once that is addressed:

- https://github.com/microsoft/hcsshim/issues/721 Revert 718 when Windows 19H1 has expand sandbox fix

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by vikramhh** [comment]
@thaJeztah @kolyshkin - does this look good to you. The test commit has been separated out and precedes the shim bump.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -3945,11 +3945,12 @@ func (s *DockerSuite) TestRunAttachFailedNoLeak(c *testing.T) {
...
-		strings.Contains(out, "were not connected because a duplicate name exists") ||
-		strings.Contains(out, "The specified port already exists") ||
-		strings.Contains(out, "HNS failed with error : Failed to create endpoint") ||
-		strings.Contains(out, "HNS failed with error : The object already exists"), fmt.Sprintf("Output: %s", out))
+	outLowerCase := strings.ToLower(out)
+	assert.Assert(c, strings.Contains(outLowerCase, "port is already allocated") ||
+		strings.Contains(outLowerCase, "were not connected because a duplicate name exists") ||
+		strings.Contains(outLowerCase, "the specified port already exists") ||
+		strings.Contains(outLowerCase, "hns failed with error : failed to create endpoint") ||
+		strings.Contains(outLowerCase, "hns failed with error : the object already exists"), fmt.Sprintf("Output: %s", out))
```
Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go:3953]: Looks like this actually was hiding a regression;

- https://github.com/moby/moby/pull/49202

Introduced in (upstream);

- https://github.com/microsoft/hcsshim/pull/573

**Pull request merged**